### ISSUE TITLE
NativeWindow.PollEvents does not perform all neccessary message wrangling

### ISSUE THREAD

**Issue by typpos** [opentk/opentk#1241]
### Description

`NativeWindow.PollEvents` does not perform all neccessary message wrangling and WinForms features like menu shortcut keys don't work.

### Repro steps

Create a C# .NET 5.0 Winforms Project. Add Menu Strip to Form, add a menu item, give it a shortcut key, eg, "F5", bind it to an action, eg, message box.

In `Program.Main(),` show the form, instantiate an `OpenTK.Windowing.Desktop.NativeWindow`, then run the following event loop:

```
var w = new NativeWindow(...)
while (!w.IsExiting) w.ProcessEvents();
```

Give input focus to the WinForm and try the shortcut key. It will not action.

### Expected behavior

The shortcut key is actioned.

### Actual behavior

Shortcut key is ignored.

### Related information 

The underlying issue seems to be that `glfwPollEvents` essentially does the typical `peekMessage - translatemessage - dispatchmessage` sequence which is ok for C++ applications. However, in WinForms, `Application.DoEvents` performs many more actions in the default WinForms message loop. There are likely a number of other WinForms / .NET features that don't work with `glfwPollEvents`.

Workaround:

Replace `NativeWindow.ProcessEvents()` with `Application.DoEvents()`. However, `NativeWindow.ProcessInputEvents` is private and cannot be called anymore, so you can't have your cake and eat it too.

Priority: Low for myself now that I know, but it was a pain to figure this out.

Win 10 Pro 10.0.19041
NVIDIA GTX 1070
OpenTK 4.5.0 (nuget)
.NET 5.0

**Comment by NogginBops** [comment]
This is not really an intended use case and is not really expected to work.
Why are you using a `NativeWindow` inside of a winforms application?

**Comment by typpos** [comment]
> Why are you using a NativeWindow inside of a winforms application?

Well, I need a window to show OpenGL rendered imagery, and another window where a user can edit GLSL source code and use traditional WinForms controls to select colours, files, vectors etc.

 Imagine the application to be a glsl source editor with live preview (though there's more to it).


> This is not really an intended use case and is not really expected to work.

Where can I find the intended use cases?

**Comment by NogginBops** [comment]
Ok I think I'm kind of getting some idea of what you are trying to create but I'm not sure I understand completely.
Can you walk me through what is using what and why?
Why is `NativeWindow` and glfw processing the event queue of the WinForms application?

**Comment by typpos** [comment]
> Can you walk me through what is using what and why?

Conceptually / simplified:

class Program:
* Orchestrates components; instantiates MyNativeWindow, EditForm, Renderer
* Runs event loop (DoEvents + Renderer.Step() + ... )
    * Step() returns info that adjusts the loop behaviour (seel below)

class MyNativeWindow : NativeWindow
* Responsible for making an OpenGL context visible on screen, handling resize, min/max, etc

class EditForm:
* Provides user with various UI / traditional means to create glsl code describing multiple glsl programs, and instructions about how they need to run (the RenderSystem).

class Renderer
* Issues GL requests based on RenderSystem in each Step() call
* A RenderSystem can require rendering to be iterative (full speed), display speed (aka 60Hz) or event driven (draw once and stop). This is taken care of by Step() returning info to the event loop which can adjust accordingly.


> Why is NativeWindow and glfw processing the event queue of the WinForms application?

Simply because I started out with an online example that used NativeWindow.ProcessEvents(). As I don't know the internal workings of OpenTK/GLFW, and because NativeWindow isn't a WinForms Form,  it wasn't a trivial step to simply abandon this "assumed required" event processing mechanism in favour of DoEvents. Plus, I now can't use the inputs that NativeWindow normally provides so I feels like I'm doing it wrong.

It's all fine for me where I am now, but it's a bit of a slog to get there...

**Comment by seanofw** [comment]
Have you considered instead using the new GLControl? This sounds like the kind of use case it was designed for.

Deep under the hood, it’s still a wrapper around a NativeWindow, but it tries to hide that fact to play nice with WinForms:  I suspect that the problems you have with the dual event loops don’t exist in it because it goes to a lot of effort to make sure that the WinForms loop is the only one receiving most of the messages.

I think you should at least give it a try in lieu of what you have; a lot of the problems you’re running into are likely already solved with it.

**Comment by typpos** [comment]
Thanks, seanofw. At the risk of distracting from this original issue report..

I started with the GLControl about 9 months ago and remember I had to rework a few minor bits in its source to make it work the way I needed it. Recently when I restarted the project, that old GLControl no longer worked with OpenTK 4.5.

I decided not to use the new GLControl because a quick scan through the code made it look at like an informal implementation that tried to force the NativeWindow to do something it wasn't designed for, and it seemed to add unwanted baggage. Please don't take this as criticism, I only did a brief review and don't have in-depth project knowledge. So, I used NativeWindow, short of just going GLFW.CreateWindow all by myself.

Example issue with GLControl:

A number of GLControl methods, eg, public property setter GLControl.API, end up calling GLControl.DestroyNativeWindow which in turn calls NativeWindow.Dispose to try to get rid of the NativeWindow instance. However, NativeWindow.Dispose does NOT call GLFW.DestroyWindow, so the actual HWND, GLFW WindowPtr stuff, etc don't get disposed. (I may be wrong).

**Comment by seanofw** [comment]
The original GLControl worked in a completely different way than the new one does; the new one didn't exist before I wrote it in December 2020.

The new one was built _from the ground up_ to have a very similar programmatic interface to the old control (and it even copies some of the same comments), but it really was _started from scratch_ and was designed to match the way that OpenTK 4 — and more specifically, GLFW — work, deep under the hood.

-----

**Quick technical deep-dive into the new code:**

Rather than attempt to force the `GLControl` itself to be the actual OpenGL window — which is what the GLControl 3.x did, and which would be fraught with all sorts of danger and wouldn't work well with GLFW — the new control reparents the underlying GLFW `HWND` as a child of the control itself, after GLFW creates it.

That's the same technique used by many [other projects](https://stackoverflow.com/questions/46152212/embed-windowglfwcreatewindow-as-child-to-c-mfc-parent-form) using GLFW in C++ to support other GUI libraries on Windows:  We're just doing it in C# instead.  The `NativeWindow` class is itself a thin wrapper around a pointer to GLFW's underlying `Window` object, which is itself a medium-sized wrapper around the actual `HWND` — so reparenting is predictable and safe (even if it looks strange), as long as you get the parent window class and the child window styles set up correctly.  This isn't guesswork; it's a pretty well-established solution at this point.

Importantly, GLFW's message pump never runs — but it doesn't need to, since WinForms's message pump does all the same things it would have, and a lot more.  And WinForms will still pump messages to the GLFW `WndProc`, so it won't starve.  (Also, GLFW's message pump's weird hacks never run; they're pretty problematic if you're handling messages elsewhere.  But if GLFW can handle being run inside the lunacy that is MFC's message processing, it can certainly handle being run inside the much cleaner message pump of WinForms!)

The only other remaining issue is that Windows (the OS) tends to be uncooperative with focus changes; so we go to some effort to ensure that GLFW never ends up receiving the input focus long enough to cause trouble, no matter how the focus is transitioned, and that ensures that input gets handled all the usual WinForms ways — unless you flip the control into "raw input mode," where it will instead force the focus away from the parent control so that GLFW can do the input processing instead, and that'll give you the data straight from the `HRAWINPUT` stream in the `WM_INPUT` events.

-----

For what it's worth, I've been coding Windows since the Windows 3.1 days, and I know a lot of the weird quirks and oddities in it pretty well.  There might be corner cases I missed in the GLControl, but I'm pretty sure all the obvious issues are covered.

In short, if you haven't used the new control, please give it a chance.  Many people are using it without trouble, and you might be able to as well.

**Comment by typpos** [comment]
Ok. That reads all very encouraging and is useful info. I think I'll have another look at the all new GLControl after my current sprint. Thanks for taking the time.

**Comment by frederikja163** [comment]
@typpos Did you manage to fix it, or do you still have issues?

**Comment by typpos** [comment]
I ended up staying with NativeWindow and my own event-loop. I didn't change over to GLControl. Itrequires some work-arounds here and there but is good enough for what I'm doing.

This issue can be marked closed, if that's the reason for asking.

**Comment by frederikja163** [comment]
It was, glad you found a work around that works for you.
